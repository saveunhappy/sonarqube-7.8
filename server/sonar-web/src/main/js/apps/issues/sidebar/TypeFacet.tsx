/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router';
import { orderBy, without } from 'lodash';
import DeferredSpinner from '../../../components/common/DeferredSpinner';
import FacetBox from '../../../components/facet/FacetBox';
import FacetHeader from '../../../components/facet/FacetHeader';
import FacetItem from '../../../components/facet/FacetItem';
import FacetItemsList from '../../../components/facet/FacetItemsList';
import HelpTooltip from '../../../components/controls/HelpTooltip';
import IssueTypeIcon from '../../../components/ui/IssueTypeIcon';
import MultipleSelectionHint from '../../../components/facet/MultipleSelectionHint';
import NewsBox from '../../../components/ui/NewsBox';
import { formatFacetStat, Query } from '../utils';
import { getCurrentUser, getCurrentUserSetting, Store } from '../../../store/rootReducer';
import { setCurrentUserSetting } from '../../../store/users';
import { ISSUE_TYPES } from '../../../helpers/constants';
import { translate } from '../../../helpers/l10n';

interface Props {
  fetching: boolean;
  newsBoxDismissHotspots?: boolean;
  onChange: (changes: Partial<Query>) => void;
  onToggle: (property: string) => void;
  open: boolean;
  setCurrentUserSetting: (setting: T.CurrentUserSetting) => void;
  stats: T.Dict<number> | undefined;
  types: string[];
}

export class TypeFacet extends React.PureComponent<Props> {
  property = 'types';

  static defaultProps = {
    open: true
  };

  handleItemClick = (itemValue: string, multiple: boolean) => {
    const { types } = this.props;
    if (multiple) {
      const newValue = orderBy(
        types.includes(itemValue) ? without(types, itemValue) : [...types, itemValue]
      );
      this.props.onChange({ [this.property]: newValue });
    } else {
      this.props.onChange({
        [this.property]: types.includes(itemValue) && types.length < 2 ? [] : [itemValue]
      });
    }
  };

  handleHeaderClick = () => {
    this.props.onToggle(this.property);
  };

  handleClear = () => {
    this.props.onChange({ [this.property]: [] });
  };

  handleDismiss = () => {
    this.props.setCurrentUserSetting({ key: 'newsbox.dismiss.hotspots', value: 'true' });
  };

  getStat(type: string) {
    const { stats } = this.props;
    return stats ? stats[type] : undefined;
  }

  isFacetItemActive(type: string) {
    return this.props.types.includes(type);
  }

  stopPropagation = (event: React.MouseEvent<HTMLAnchorElement>) => {
    event.stopPropagation();
  };

  renderItem = (type: string) => {
    const active = this.isFacetItemActive(type);
    const stat = this.getStat(type);

    return (
      <FacetItem
        active={active}
        disabled={stat === 0 && !active}
        key={type}
        name={
          <span className="display-flex-center">
            <IssueTypeIcon className="little-spacer-right" query={type} />{' '}
            {translate('issue.type', type)}
            {type === 'SECURITY_HOTSPOT' && this.props.newsBoxDismissHotspots && (
              <HelpTooltip
                className="little-spacer-left"
                overlay={
                  <>
                    <p>{translate('issues.hotspots.helper')}</p>
                    <hr className="spacer-top spacer-bottom" />
                    <Link
                      onClick={this.stopPropagation}
                      target="_blank"
                      to="/documentation/user-guide/security-hotspots/">
                      {translate('learn_more')}
                    </Link>
                  </>
                }
              />
            )}
          </span>
        }
        onClick={this.handleItemClick}
        stat={formatFacetStat(stat)}
        value={type}
      />
    );
  };

  render() {
    const { newsBoxDismissHotspots, types, stats = {} } = this.props;
    const values = types.map(type => translate('issue.type', type));

    const showHotspotNewsBox =
      types.includes('SECURITY_HOTSPOT') || (types.length === 0 && stats['SECURITY_HOTSPOT'] > 0);

    return (
      <FacetBox property={this.property}>
        <FacetHeader
          clearLabel="reset_verb"
          name={translate('issues.facet', this.property)}
          onClear={this.handleClear}
          onClick={this.handleHeaderClick}
          open={this.props.open}
          values={values}
        />

        <DeferredSpinner loading={this.props.fetching} />
        {this.props.open && (
          <>
            <FacetItemsList>{ISSUE_TYPES.map(this.renderItem)}</FacetItemsList>
            {!newsBoxDismissHotspots && showHotspotNewsBox && (
              <NewsBox
                onClose={this.handleDismiss}
                title={translate('issue.type.SECURITY_HOTSPOT.plural')}>
                <p>{translate('issues.hotspots.helper')}</p>
                <p className="text-right spacer-top">
                  <Link target="_blank" to="/documentation/user-guide/security-hotspots/">
                    {translate('learn_more')}
                  </Link>
                </p>
              </NewsBox>
            )}
            <MultipleSelectionHint options={Object.keys(stats).length} values={types.length} />
          </>
        )}
      </FacetBox>
    );
  }
}

const mapStateToProps = (state: Store) => ({
  newsBoxDismissHotspots:
    !getCurrentUser(state).isLoggedIn ||
    getCurrentUserSetting(state, 'newsbox.dismiss.hotspots') === 'true'
});

const mapDispatchToProps = {
  setCurrentUserSetting
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(TypeFacet);
