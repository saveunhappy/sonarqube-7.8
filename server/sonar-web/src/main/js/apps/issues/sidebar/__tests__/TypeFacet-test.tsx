/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import { TypeFacet } from '../TypeFacet';
import { click } from '../../../../helpers/testUtils';

it('should render open by default', () => {
  expect(shallowRender({ types: ['VULNERABILITY', 'CODE_SMELL'] })).toMatchSnapshot();
});

it('should toggle type facet', () => {
  const onToggle = jest.fn();
  const wrapper = shallowRender({ onToggle });
  click(wrapper.children('FacetHeader'));
  expect(onToggle).toBeCalledWith('types');
});

it('should clear types facet', () => {
  const onChange = jest.fn();
  const wrapper = shallowRender({ onChange, types: ['BUGS'] });
  wrapper.children('FacetHeader').prop<Function>('onClear')();
  expect(onChange).toBeCalledWith({ types: [] });
});

it('should select a type', () => {
  const onChange = jest.fn();
  const wrapper = shallowRender({ onChange });
  clickAndCheck('CODE_SMELL');
  clickAndCheck('VULNERABILITY', true, ['CODE_SMELL', 'VULNERABILITY']);
  clickAndCheck('SECURITY_HOTSPOT');

  function clickAndCheck(type: string, multiple = false, expected = [type]) {
    wrapper
      .find(`FacetItemsList`)
      .find(`FacetItem[value="${type}"]`)
      .prop<Function>('onClick')(type, multiple);
    expect(onChange).lastCalledWith({ types: expected });
    wrapper.setProps({ types: expected });
  }
});

it('should display the hotspot newsbox', () => {
  expect(shallowRender({ types: ['SECURITY_HOTSPOT'] }).find('NewsBox')).toMatchSnapshot();
  expect(
    shallowRender({ types: [] })
      .find('NewsBox')
      .exists()
  ).toBe(true);
});

it('should display the hotspot tooltip helper only', () => {
  let wrapper = shallowRender({ types: ['SECURITY_HOTSPOT'], newsBoxDismissHotspots: true });
  expect(wrapper.find('NewsBox').exists()).toBe(false);
  expect(
    wrapper
      .find(`FacetItemsList`)
      .find(`FacetItem[value="SECURITY_HOTSPOT"]`)
      .prop('name')
  ).toMatchSnapshot();

  wrapper = shallowRender({ types: ['BUGS'], newsBoxDismissHotspots: true });
  expect(wrapper.find('NewsBox').exists()).toBe(false);
  expect(
    wrapper
      .find(`FacetItemsList`)
      .find(`FacetItem[value="SECURITY_HOTSPOT"]`)
      .prop('name')
  ).toMatchSnapshot();
});

function shallowRender(props: Partial<TypeFacet['props']> = {}) {
  return shallow(
    <TypeFacet
      fetching={false}
      onChange={jest.fn()}
      onToggle={jest.fn()}
      setCurrentUserSetting={jest.fn()}
      stats={{ BUG: 0, VULNERABILITY: 2, CODE_SMELL: 5, SECURITY_HOTSPOT: 1 }}
      types={[]}
      {...props}
    />
  );
}
