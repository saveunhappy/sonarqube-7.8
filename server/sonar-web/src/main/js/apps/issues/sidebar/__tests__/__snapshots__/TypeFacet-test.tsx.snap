// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display the hotspot newsbox 1`] = `
<NewsBox
  onClose={[Function]}
  title="issue.type.SECURITY_HOTSPOT.plural"
>
  <p>
    issues.hotspots.helper
  </p>
  <p
    className="text-right spacer-top"
  >
    <Link
      onlyActiveOnIndex={false}
      style={Object {}}
      target="_blank"
      to="/documentation/user-guide/security-hotspots/"
    >
      learn_more
    </Link>
  </p>
</NewsBox>
`;

exports[`should display the hotspot tooltip helper only 1`] = `
<span
  className="display-flex-center"
>
  <IssueTypeIcon
    className="little-spacer-right"
    query="SECURITY_HOTSPOT"
  />
   
  issue.type.SECURITY_HOTSPOT
  <HelpTooltip
    className="little-spacer-left"
    overlay={
      <React.Fragment>
        <p>
          issues.hotspots.helper
        </p>
        <hr
          className="spacer-top spacer-bottom"
        />
        <Link
          onClick={[Function]}
          onlyActiveOnIndex={false}
          style={Object {}}
          target="_blank"
          to="/documentation/user-guide/security-hotspots/"
        >
          learn_more
        </Link>
      </React.Fragment>
    }
  />
</span>
`;

exports[`should display the hotspot tooltip helper only 2`] = `
<span
  className="display-flex-center"
>
  <IssueTypeIcon
    className="little-spacer-right"
    query="SECURITY_HOTSPOT"
  />
   
  issue.type.SECURITY_HOTSPOT
  <HelpTooltip
    className="little-spacer-left"
    overlay={
      <React.Fragment>
        <p>
          issues.hotspots.helper
        </p>
        <hr
          className="spacer-top spacer-bottom"
        />
        <Link
          onClick={[Function]}
          onlyActiveOnIndex={false}
          style={Object {}}
          target="_blank"
          to="/documentation/user-guide/security-hotspots/"
        >
          learn_more
        </Link>
      </React.Fragment>
    }
  />
</span>
`;

exports[`should render open by default 1`] = `
<FacetBox
  property="types"
>
  <FacetHeader
    clearLabel="reset_verb"
    name="issues.facet.types"
    onClear={[Function]}
    onClick={[Function]}
    open={true}
    values={
      Array [
        "issue.type.VULNERABILITY",
        "issue.type.CODE_SMELL",
      ]
    }
  />
  <DeferredSpinner
    loading={false}
    timeout={100}
  />
  <FacetItemsList>
    <FacetItem
      active={false}
      disabled={true}
      halfWidth={false}
      key="BUG"
      loading={false}
      name={
        <span
          className="display-flex-center"
        >
          <IssueTypeIcon
            className="little-spacer-right"
            query="BUG"
          />
           
          issue.type.BUG
        </span>
      }
      onClick={[Function]}
      stat={0}
      value="BUG"
    />
    <FacetItem
      active={true}
      disabled={false}
      halfWidth={false}
      key="VULNERABILITY"
      loading={false}
      name={
        <span
          className="display-flex-center"
        >
          <IssueTypeIcon
            className="little-spacer-right"
            query="VULNERABILITY"
          />
           
          issue.type.VULNERABILITY
        </span>
      }
      onClick={[Function]}
      stat="2"
      value="VULNERABILITY"
    />
    <FacetItem
      active={true}
      disabled={false}
      halfWidth={false}
      key="CODE_SMELL"
      loading={false}
      name={
        <span
          className="display-flex-center"
        >
          <IssueTypeIcon
            className="little-spacer-right"
            query="CODE_SMELL"
          />
           
          issue.type.CODE_SMELL
        </span>
      }
      onClick={[Function]}
      stat="5"
      value="CODE_SMELL"
    />
    <FacetItem
      active={false}
      disabled={false}
      halfWidth={false}
      key="SECURITY_HOTSPOT"
      loading={false}
      name={
        <span
          className="display-flex-center"
        >
          <IssueTypeIcon
            className="little-spacer-right"
            query="SECURITY_HOTSPOT"
          />
           
          issue.type.SECURITY_HOTSPOT
        </span>
      }
      onClick={[Function]}
      stat="1"
      value="SECURITY_HOTSPOT"
    />
  </FacetItemsList>
  <MultipleSelectionHint
    options={4}
    values={2}
  />
</FacetBox>
`;
